File: app/src/main/java/com/easy/kotlins/http/core/OkCallbacks.java
package com.easy.kotlins.http.core;

import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.util.Log;

import com.facebook.stetho.BuildConfig;

import java.util.Objects;

import okhttp3.Response;

/**
 * Create by LiZhanPing on 2020/4/27
 * desc: 请求回调处理
 */
final class OkCallbacks {

    private static final int MSG_WHAT_BASE = 1000000000;
    private static final int MSG_WHAT_ON_START = MSG_WHAT_BASE + 1;
    private static final int MSG_WHAT_ON_SUCCESS = MSG_WHAT_BASE + 2;
    private static final int MSG_WHAT_ON_ERROR = MSG_WHAT_BASE + 3;
    private static final int MSG_WHAT_ON_UPDATE = MSG_WHAT_BASE + 4;
    private static final int MSG_WHAT_ON_CANCEL = MSG_WHAT_BASE + 5;

    static final class MessageBody {
        final OkCallback<?> callback;
        final Object[] args;

        private MessageBody(OkCallback<?> callback, Object... args) {
            this.callback = callback;
            this.args = args;
        }
    }

    static final Handler HANDLER = new Handler(Looper.getMainLooper()) {

        @Override
        public void handleMessage(Message msg) {
            MessageBody body = (MessageBody) msg.obj;
            switch (msg.what) {
                case MSG_WHAT_ON_START: {
                    @SuppressWarnings("unchecked") OkStartedCallback<Object> callback = (OkStartedCallback<Object>) body.callback;
                    callOnStart(callback);
                }
                break;
                case MSG_WHAT_ON_SUCCESS: {
                    @SuppressWarnings("unchecked") OkCallback<Object> callback = (OkCallback<Object>) body.callback;
                    callOnSuccess(callback, body.args[0]);
                }
                break;
                case MSG_WHAT_ON_ERROR: {
                    callOnError(body.callback, (Throwable) body.args[0]);
                }
                break;
                case MSG_WHAT_ON_UPDATE: {
                    callOnProgress((OkDownloadCallback) body.callback, (long) body.args[0], (long) body.args[1]);
                }
                break;
                case MSG_WHAT_ON_CANCEL: {
                    callOnCancel((OkDownloadCallback) body.callback);
                }
                break;
            }
        }
    };

    static <T> void response(final OkGenericCallback<T> callback, final Response response) {
        if (callback != null) {
            T result = null;
            Exception failure = null;
            try {
                result = callback.parseResponse(Objects.requireNonNull(response.body()).string());

                if (result == null) {
                    throw new NullPointerException("result is null");
                }
            } catch (Exception e) {
                failure = e;
            }

            if (failure != null) {
                HANDLER.obtainMessage(MSG_WHAT_ON_ERROR, new MessageBody(callback, failure)).sendToTarget();
            } else {
                HANDLER.obtainMessage(MSG_WHAT_ON_SUCCESS, new MessageBody(callback, result)).sendToTarget();
            }
        }
    }

    static <T> void success(final OkCallback<T> callback, final T result) {
        if (callback != null) {
            HANDLER.obtainMessage(MSG_WHAT_ON_SUCCESS, new MessageBody(callback, result)).sendToTarget();
        }
    }

    static void error(final OkCallback<?> callback, final Throwable error) {
        if (callback != null) {
            HANDLER.obtainMessage(MSG_WHAT_ON_ERROR, new MessageBody(callback, error)).sendToTarget();
        }
    }

    static void start(final OkStartedCallback<?> callback) {
        if (callback != null) {
            HANDLER.obtainMessage(MSG_WHAT_ON_START, new MessageBody(callback)).sendToTarget();
        }
    }

    static void cancel(final OkDownloadCallback callback) {
        if (callback != null) {
            HANDLER.obtainMessage(MSG_WHAT_ON_CANCEL, new MessageBody(callback)).sendToTarget();
        }
    }

    static void progress(final OkDownloadCallback callback, final long downloadedBytes, final long totalBytes) {
        if (callback != null) {
            HANDLER.obtainMessage(MSG_WHAT_ON_UPDATE, new MessageBody(callback, downloadedBytes, totalBytes)).sendToTarget();
        }
    }

    private static void callOnStart(final OkStartedCallback<?> callback) {
        try {
            callback.onStart();
        } catch (Throwable t) {
            t.printStackTrace();
        }
    }

    private static void callOnCancel(final OkDownloadCallback callback) {
        try {
            callback.onCancel();
        } catch (Throwable t) {
            t.printStackTrace();
        }
    }

    private static void callOnProgress(final OkDownloadCallback callback, long downloadedBytes, long totalBytes) {
        try {
            callback.onProgress(downloadedBytes, totalBytes);
        } catch (Throwable t) {
            t.printStackTrace();
        }
    }

    private static <T> void callOnSuccess(final OkCallback<T> callback, final T result) {
        try {
            callback.onSuccess(result);
        } catch (Exception e) {
            callOnError(callback, e);
        }
    }

    private static void callOnError(final OkCallback<?> callback, Throwable e) {
        try {
            if (BuildConfig.DEBUG) {
                Log.e("OkFaker", e.getLocalizedMessage(), e);
            }

            callback.onError(e);
        } catch (Throwable t) {
            t.printStackTrace();
        }
    }

}

File: common-ktx/src/main/java/com/easy/kotlins/http/core/OkCallbacks.kt
package com.easy.kotlins.http.core

import android.os.Handler
import android.os.Looper
import android.os.Message
import android.util.Log
import com.facebook.stetho.BuildConfig

/**
 * Create by LiZhanPing on 2020/4/27
 * desc: 请求回调处理
 */
internal object OkCallbacks {
    private const val MSG_WHAT_BASE = 1000000000
    private const val MSG_WHAT_ON_START = MSG_WHAT_BASE + 1
    private const val MSG_WHAT_ON_SUCCESS = MSG_WHAT_BASE + 2
    private const val MSG_WHAT_ON_ERROR = MSG_WHAT_BASE + 3
    private const val MSG_WHAT_ON_UPDATE = MSG_WHAT_BASE + 4
    private const val MSG_WHAT_ON_CANCEL = MSG_WHAT_BASE + 5
    private const val MSG_WHAT_ON_COMPLETE = MSG_WHAT_BASE + 6
    private val HANDLER: Handler = object : Handler(Looper.getMainLooper()) {
        override fun handleMessage(msg: Message) {
            val body = msg.obj as MessageBody
            when (msg.what) {
                MSG_WHAT_ON_START -> {
                    @Suppress("UNCHECKED_CAST") val callback = body.callback as OkCallback<Any>
                    callOnStart(callback)
                }
                MSG_WHAT_ON_COMPLETE -> {
                    @Suppress("UNCHECKED_CAST") val callback = body.callback as OkCallback<Any>
                    callOnComplete(callback)
                }
                MSG_WHAT_ON_SUCCESS -> {
                    @Suppress("UNCHECKED_CAST") val callback = body.callback as OkCallback<Any>
                    callOnSuccess(callback, body.args[0])
                }
                MSG_WHAT_ON_ERROR -> {
                    callOnError(body.callback, body.args[0] as Throwable)
                }
                MSG_WHAT_ON_UPDATE -> {
                    callOnProgress(
                        body.callback as OkDownloadCallback,
                        body.args[0] as Long,
                        body.args[1] as Long
                    )
                }
                MSG_WHAT_ON_CANCEL -> {
                    callOnCancel(body.callback as OkDownloadCallback)
                }
            }
        }
    }

    fun <T: Any> success(callback: OkCallback<T>?, result: T) {
        if (callback != null) {
            HANDLER.obtainMessage(MSG_WHAT_ON_SUCCESS, MessageBody(callback, result)).sendToTarget()
        }
    }

    fun error(callback: OkCallback<*>?, error: Throwable?) {
        if (callback != null) {
            HANDLER.obtainMessage(MSG_WHAT_ON_ERROR, MessageBody(callback, error!!)).sendToTarget()
        }
    }

    fun start(callback: OkCallback<*>?) {
        if (callback != null) {
            HANDLER.obtainMessage(MSG_WHAT_ON_START, MessageBody(callback)).sendToTarget()
        }
    }

    fun complete(callback: OkCallback<*>?) {
        if (callback != null) {
            HANDLER.obtainMessage(MSG_WHAT_ON_COMPLETE, MessageBody(callback)).sendToTarget()
        }
    }

    fun cancel(callback: OkDownloadCallback?) {
        if (callback != null) {
            HANDLER.obtainMessage(MSG_WHAT_ON_CANCEL, MessageBody(callback)).sendToTarget()
        }
    }

    fun progress(callback: OkDownloadCallback?, downloadedBytes: Long, totalBytes: Long) {
        if (callback != null) {
            HANDLER.obtainMessage(
                MSG_WHAT_ON_UPDATE,
                MessageBody(callback, downloadedBytes, totalBytes)
            ).sendToTarget()
        }
    }

    private fun callOnStart(callback: OkCallback<*>) {
        try {
            callback.onStart()
        } catch (t: Throwable) {
            t.printStackTrace()
        }
    }

    private fun callOnComplete(callback: OkCallback<*>) {
        try {
            callback.onComplete()
        } catch (t: Throwable) {
            t.printStackTrace()
        }
    }

    private fun callOnCancel(callback: OkDownloadCallback) {
        try {
            callback.onCancel()
        } catch (t: Throwable) {
            t.printStackTrace()
        }
    }

    private fun callOnProgress(
        callback: OkDownloadCallback,
        downloadedBytes: Long,
        totalBytes: Long
    ) {
        try {
            callback.onProgress(downloadedBytes, totalBytes)
        } catch (t: Throwable) {
            t.printStackTrace()
        }
    }

    private fun <T> callOnSuccess(callback: OkCallback<T>, result: T) {
        try {
            callback.onSuccess(result)
        } catch (e: Exception) {
            callOnError(callback, e)
        }
    }

    private fun callOnError(callback: OkCallback<*>, e: Throwable) {
        try {
            if (BuildConfig.DEBUG) {
                Log.e(OkRequest::class.simpleName, e.localizedMessage, e)
            }
            callback.onError(e)
        } catch (t: Throwable) {
            t.printStackTrace()
        }
    }

    internal class MessageBody(val callback: OkCallback<*>, vararg args: Any) {
        val args: Array<Any> = arrayOf(args)
    }
}